sdks/python/apache_beam/runners/worker/opcounters.py

```diff
@@ -24,11 +24,117 @@
 import math
 import random
 
+from apache_beam.utils import counters
 from apache_beam.utils.counters import Counter
+from apache_beam.utils.counters import CounterName
 
 # This module is experimental. No backwards-compatibility guarantees.
 
 
+class TransformIOCounter(object):
+  """Class to track time and bytes consumed while reading from IO.
+
+  Subclasses should be able to track consumption of IO across steps
+  in the same stage - for instance, if a Shuffle or Side Input iterable
+  is passed down to a next step.
+
+  Some examples of IO can be side inputs, shuffle, or streaming state.
+  """
+
+  def update_current_step(self):
+    """Update the current step within a stage as it may have changed.
+
+    If the state changed, it would mean that an initial step passed a
+    data-accessor (such as a side input / shuffle Iterable) down to the
+    next step in a stage.
+    """
+    pass
+
+  def add_bytes_read(self, count):
+    pass
+
+  def __exit__(self, exc_type, exc_value, traceback):
+    """Exit the IO state."""
+    pass
+
+  def __enter__(self):
+    """Enter the IO state. This should track time spent blocked on IO."""
+    pass
+
+
+class SideInputReadCounter(TransformIOCounter):
+  """Tracks time and bytes consumed while reading from side inputs.
+
+  This class is designed to track consumption of side inputs across fused steps.
+  We represent a side input as a declaring step, and an input index, which are
+  represented as an io_target.
```

Comment:
<!--thread_id:cc_163387836_t; commit:ac6abcff9e381c9345f7aaeabdf8f83998f6074c; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> I'm a little confused because this sentence seems to imply that the input index is an object of type `io_target`, which seems to clash with the documentation in `__init__` below (which is otherwise very clear).

<!--section:body-->
Hmm I see the point. Yes, io_target is only used later in the CounterName. I've decided to document it there, and not mention it here.